pkg/resources/resource_source_mysql.go

```diff
@@ -0,0 +1,168 @@
+					Description: "An alias for the table, used in Materialize.",
+					Type:        schema.TypeString,
+					Optional:    true,
+				},
+			},
+		},
+	},
+	"subsource":      SubsourceSchema(),
+	"ownership_role": OwnershipRoleSchema(),
+	"region":         RegionSchema(),
+}
+
+func SourceMySQL() *schema.Resource {
+	return &schema.Resource{
+		Description: "A MySQL source describes a MySQL instance you want Materialize to read data from.",
+
+		CreateContext: sourceMySQLCreate,
+		ReadContext:   sourceRead,
+		UpdateContext: sourceMySQLUpdate,
+		DeleteContext: sourceDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: sourceMySQLSchema,
+	}
+}
+
+func sourceMySQLCreate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	sourceName := d.Get("name").(string)
+	schemaName := d.Get("schema_name").(string)
+	databaseName := d.Get("database_name").(string)
+
+	metaDb, region, err := utils.GetDBClientFromMeta(meta, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	o := materialize.MaterializeObject{ObjectType: "SOURCE", Name: sourceName, SchemaName: schemaName, DatabaseName: databaseName}
+	b := materialize.NewSourceMySQLBuilder(metaDb, o)
+
+	if v, ok := d.GetOk("cluster_name"); ok {
+		b.ClusterName(v.(string))
+	}
+
+	if v, ok := d.GetOk("mysql_connection"); ok {
+		conn := materialize.GetIdentifierSchemaStruct(v)
+		b.MySQLConnection(conn)
+	}
+
+	if v, ok := d.GetOk("table"); ok {
+		tables := materialize.GetTableStruct(v.([]interface{}))
+		b.Tables(tables)
+	}
+
+	if err := b.Create(); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Handle ownership
+	if v, ok := d.GetOk("ownership_role"); ok {
+		ownership := materialize.NewOwnershipBuilder(metaDb, o)
+
+		if err := ownership.Alter(v.(string)); err != nil {
+			log.Printf("[DEBUG] resource failed ownership, dropping object: %s", o.Name)
+			b.Drop()
+			return diag.FromErr(err)
+		}
+	}
+
+	// Handle comments
+	if v, ok := d.GetOk("comment"); ok {
+		comment := materialize.NewCommentBuilder(metaDb, o)
+
+		if err := comment.Object(v.(string)); err != nil {
+			log.Printf("[DEBUG] resource failed comment, dropping object: %s", o.Name)
+			b.Drop()
+			return diag.FromErr(err)
+		}
+	}
+
+	i, err := materialize.SourceId(metaDb, o)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(utils.TransformIdWithRegion(string(region), i))
+
+	return sourceRead(ctx, d, meta)
+}
+
+func sourceMySQLUpdate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	sourceName := d.Get("name").(string)
+	schemaName := d.Get("schema_name").(string)
+	databaseName := d.Get("database_name").(string)
+
+	metaDb, _, err := utils.GetDBClientFromMeta(meta, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	o := materialize.MaterializeObject{ObjectType: "SOURCE", Name: sourceName, SchemaName: schemaName, DatabaseName: databaseName}
+	// b := materialize.NewSource(metaDb, o)
```

Comment:
is this relevant? should we allow updating a source object at all?

Reply:
Yes, you can update some propertites like the the name of the source, comments, ownership and etc.

I still have to work on implementing the subsources as well, eg. adding and removing subsources `ALTER SOURCE "database"."schema"."source" ADD SUBSOURCE "table_1" AS "table_alias";`

Reply:
Ah got it! Just so you know the `ALTER SOURCE` statements are not yet supported for MySQL.

Reply:
Ah yes, I've created a follow up issue for this: https://github.com/MaterializeInc/terraform-provider-materialize/issues/490